fastfield_codecs/src/column.rs

```diff
@@ -58,10 +66,70 @@ pub trait Column<T: PartialOrd = u64>: Send + Sync {
     fn max_value(&self) -> T;
 
     fn num_vals(&self) -> u64;
+}
+
+/// `ColumnReader` makes it possible to read forward through a column.
+pub trait ColumnReader<T = u64> {
+    /// Advance the reader to the target_idx.
+    ///
+    /// After a successful call to seek,
+    /// `.get()` should returns `column.get_val(target_idx)`.
+    fn seek(&mut self, target_idx: u64) -> T;
 
-    /// Returns a iterator over the data
-    fn iter<'a>(&'a self) -> Box<dyn Iterator<Item = T> + 'a> {
-        Box::new((0..self.num_vals()).map(|idx| self.get_val(idx)))
+    fn advance(&mut self) -> bool;
+
+    /// Get the current value without advancing the reader
+    fn get(&self) -> T;
+}
+
+pub fn iter_from_reader<'a, T: 'static>(
+    mut column_reader: Box<dyn ColumnReader<T> + 'a>,
+) -> impl Iterator<Item = T> + 'a {
+    std::iter::from_fn(move || {
+        if !column_reader.advance() {
+            return None;
+        }
+        Some(column_reader.get())
+    })
+}
+
+pub(crate) struct ColumnReaderAdapter<'a, C: ?Sized, T> {
+    column: &'a C,
+    idx: u64,
+    len: u64,
+    _phantom: PhantomData<T>,
+}
+
+impl<'a, C: Column<T> + ?Sized, T: Copy + PartialOrd + 'static> From<&'a C>
+    for ColumnReaderAdapter<'a, C, T>
+{
+    fn from(column: &'a C) -> Self {
+        ColumnReaderAdapter {
+            column,
+            idx: u64::MAX,
+            len: column.num_vals(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a, T, C: ?Sized> ColumnReader<T> for ColumnReaderAdapter<'a, C, T>
+where
+    C: Column<T>,
+    T: PartialOrd<T> + Copy + 'static,
+{
+    fn seek(&mut self, idx: u64) -> T {
+        self.idx = idx;
+        self.get()
+    }
+
+    fn advance(&mut self) -> bool {
+        self.idx = self.idx.wrapping_add(1);
+        self.idx < self.len
+    }
+
+    fn get(&self) -> T {
+        self.column.get_val(self.idx)
```

Comment:
Isn't the main reason for `ColumnReader` to remove `get_val`? Or is this just a temp hack?